### ISSUE TITLE
Understanding YOLOv8 output from the Detect head in order to do an implementation from scratch in Keras

### ISSUE THREAD

**Issue by MagiPrince** [ultralytics/ultralytics#2951]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone,

For my master thesis, I am doing an implementation from scratch of YOLOv8 in Keras in order to quantize it later with QKeras (and do some modifications if necessary) for a FPGA implementation. The project aims to detect a specific element in images of 64x64, so I am dealing with only one class. 

I already took a look at the model structure and implemented successfully all the used blocks. But I am struggling to understand  the output of the Detect block for training. Actually if I take the results of the bbox part, I get three tensors which size is [batch, channels, height, width] : [1, 64, 8, 8], [1, 64, 4, 4], [1, 64, 2, 2]. I am not sure how to interpretate them... What do  (8,8), (4,4), (2,2) height and width correspond to ? And why do they always have 64 channels (that are not related to the size of the image) ? Does the (height, width) correspond to "sub-slices" of the images and the channels to the number of elements that can be detected in those "sub-slices" ? And if so, does the values of the bbox part correspond to the left top and right bottom part that may be chained for a height and width bigger than 2 ?

I will appreciate your help ! :smiley:

**Comment by MagiPrince** [comment]
I found this #https://github.com/ultralytics/ultralytics/issues/2670#issuecomment-1552627350 answer, that explains some elements, but I still don't get why there is a difference between the "indices of the mask", isn't that value fixed ?

**Comment by glenn-jocher** [comment]
@MagiPrince, the size of each  detection prediction tensor corresponds to the number of anchor boxes used during training, their aspect ratio and their scale. Each of these tensors can be seen as a feature map with a specific spatial resolution (8, 4, and 2 respectively, in YOLOv8). Their channels represent the predicted values for each anchor box at each position of this feature map: 4 boxes at each position x y and each anchor size. 

Masked values indicate which anchor boxes of the 3 anchor scales are relevant at each spatial resolution layer. Each tensor slice corresponding to one anchor scale has 3 values (width, height, objectness score) per anchor box at each spatial position. Additional values are added for each class that you are trying to predict. This means that the 3 values are expecting to predict a bounding box for each anchor box with respect to the number of classes to predict.

In summary, the indices of the mask represent which of the anchor scales and anchor boxes in the resulting tensor are used to predict bounding boxes and confidence scores for each object in the image.

**Comment by MagiPrince** [comment]
@glenn-jocher, thanks for your answer ! It's still a bit confuse to me. May I rephrase and use an example to see if I understood it well and ask for more information ?

So we have the following format : CxHxW, C (channels) correspond to the anchor boxes (this value is set in the "detect" module with 4*reg_max), H (height) correspond to the aspect ratio of the anchor boxes, and W (width) correspond to the scale of the anchor boxes.
As you said, those tensors can be seen as feature maps with a specific spatial resolution (8, 4, and 2) that in my case as I am working with images of size 64x64 results on pixels of 8, 16, and 32 in this "feature map" is that right ?
But I don't get the part of your answer : "4 boxes at each position x y and each anchor size". Can you maybe reprhrase it ?

Now, if I take an example with fictionnal values for my tensor of size 64x2x2 (output of the Sequential cv2) : 
```
tensor([[[[-8.1146e-02,  3.0194e-01],
          [-1.2012e-01,  4.7966e-01]],

         [[ 1.4206e-01,  3.1587e-01],
          [ 8.2627e-02,  4.7909e-01]],
...
```
Does it makes sens to read it like : 
```
tensor([[[[aspect ratio,  scale],
          [aspect ratio,  scale]],

         [[aspect ratio,  scale],
          [ aspect ratio,  scale]],
...
```
or like : 
```
tensor([[[[x,  y],
          [x,  y]],

         [[x,  y],
          [ x,  y]],
...
```
or in some other way ?

And as I am trying to predict only 1 class, the output of the Sequential cv3 is of size [channels, height, width]  : [1x8x8], [1x4x4], [1x2x2], where the channels correspond to the number of classes I try to dectect. But I'm not sure how to interpretate the result.
How do I know to which anchor box the result it's linked ? Or maybe at this stage I can't know it ?

Finally, which output is used for training the YOLOv8 model ? The cv2 and cv3 output since in the code those are the values returned `if self.training: return x`, or may the values (y, x) generated from the previous ones be used ?

I appreciate your help !

**Comment by MagiPrince** [comment]
Hi @glenn-jocher, thanks for your quick answer ! With this additional information it begins to make more sens. To be sure, the tensor's format that you clarify `[batch_size, num_anchors * (5 + num_classes), height, width]` concerns the cv2 output is that right ? If so, the channel size it's still a bit fuzzy for me. In the model diagram the output channels of cv2 are defined as `c=4*reg_max`, what is the link between `num_anchors * (5 + num_classes)` and `c=4*reg_max` ?

Concerning the training, have you some documentation about how the tensors are used to compute the loss for the bounding box predictions and for the objectness score and class prediction ? Indeed, I'm trying to create my function to train my custom model (written from scratch), but I'm struggling to understand how it's done in the official code...

**Comment by glenn-jocher** [comment]
Hi @MagiPrince, I'm glad to hear that my previous answer was helpful. Regarding your first question, the `[batch_size, num_anchors * (5 + num_class), height, width]` tensor format does indeed correspond to the cv2 output.

To clarify the link between `num_anchors * (5 + num_classes)` and `c=4*reg_max`, the value of `num_anchors` is equal to `c/4` since each anchor box is defined by 4 parameters (x, y, width, height). The factor of 5 in the tensor formula comes from the objectness score and the 4 parameters of each anchor box. The remaining channels are used for the class predictions, with `num_classes` channels per anchor box.

Regarding your question about how the tensors are used to compute the loss for the bounding box predictions and objectness/class predictions, the loss calculation is based on the difference between the predicted values and the ground truth values for each parameter. For the bounding box predictions, the loss is typically computed using a mean-squared error (MSE) or smooth L1 loss function. For the objectness score and class predictions, the loss is computed using binary cross-entropy or categorical cross-entropy loss, respectively.

I recommend looking at the code for the `compute_loss` function in the YOLOv8 codebase, which contains the implementation of the loss calculation. This should provide some insight into how the tensors are used to compute the loss.

I hope this helps! Let me know if you have any more questions.

**Comment by MagiPrince** [comment]
Hi @glenn-jocher, thank for your answers ! I understand the model better each time ! :smile: 

I'm still not sure to fully understand the link between `num_anchors * (5 + num_classes)` and `c=4*reg_max`. Because, the output channels of `cv2` is defined in the code (and diagram) as `4*reg_max` where `reg_max = 16`. So, we have 64 channels as output. But if we take the output size you talked about and applied it to my case where I have only 1 class, we have `num_anchors*6`. The problem here is that from `num_anchors*6`, we cannot get the 64 channels as outputed by `cv2`... Futhermore, the value of `num_class` isn't contained in `cv3` before the concatenation between `cv2` and `cv3` ?

I think, I'm missing something, because the size of the output confuse me. `num_anchors * (5 + num_classes)` reminds me of the output in YOLOv5, but I can't figure how it applies to YOLOv8 without changing the output size...

Concerning the loss computation, it's very clear thank you ! In addition to that, to know which anchor box is linked to which ground truth TaskAlignedAssigner is applied and I should compute the loss between my anchor box and the ground truth having the highest score is that right ?

Thanks for your time, you help me a lot !

**Comment by CogSP** [comment]
Hi everyone, 

I’m encountering the same issue as @MagiPrince. I’m working on recreating YOLOv8 from scratch for a university project.

I've managed to replicate the architecture, as outlined [here](https://abintimilsina.medium.com/yolov8-architecture-explained-a5e90a560ce5). Each head detector returns two tensors, one for the Bbox and one for Cls. 
![image](https://github.com/user-attachments/assets/9bbbb568-3240-4f99-83f2-e0b7101ec3a6)

Here are the shape of the two tensors:
- Bbox: [`batch_size`, `4*reg_max`, 46, 46] 
- Cls: [`batch_size`, `number_of_classes`, 46, 46]

In this context, I’m using 46 as a placeholder, but I’m not certain what it represents.

The network has three detectors in the head, each specialized for detecting objects of different sizes (large, medium, and small), as shown in this image:
![image](https://github.com/user-attachments/assets/009dab7e-a0ee-458a-b74b-3af02302bfc9)
Each of these detector is specialized for large, medium or small size objects.

Using `reg_max `= 16, `batch_size `= 1 and `number_of_classes `= 1, these are the shapes of my outputs:

- Detector 1: Bbox: [1, 64, 46, 46], Cls: [1, 1, 46, 46]
- Detector 2: Bbox: [1, 64, 22, 22], Cls: [1, 1, 22, 22]
- Detector 3: Bbox: [1, 64, 10, 10], Cls: [1, 1, 10, 10]


I’m trying to understand how to interpret these outputs, particularly in terms of how they relate to my labels, which are in the format [image_id, x_min, x_max, y_min, y_max]. How should I compare the model's outputs, shaped as I described above, with my labels in order to compute the loss?

Moreover, I’m also confused about the frequent references to anchors in discussions, especially since YOLOv8 is described as an anchor-free model. Could someone clarify why anchors are still being mentioned?

**Comment by ruti1090** [comment]
אני רוצה להבין מה אומרים לי הערכים שחוזרים מיולו

**Comment by glenn-jocher** [comment]
Hi! For YOLOv8 outputs, the Bbox tensor's 64 channels (4*reg_max=16) represent distribution bins for bounding box coordinates, not anchors. YOLOv8 is anchor-free - spatial dimensions like 46x46 correspond to grid cells at different scales for object detection. The [Model Documentation](https://docs.ultralytics.com/models/yolov8) explains how outputs map to predictions through the TaskAlignedAssigner which matches predictions to ground truth based on classification scores and IoU. For loss computation, see the `v8DetectionLoss` implementation in the [Loss Documentation](https://docs.ultralytics.com/reference/utils/loss). Let us know if you need specific implementation details! 🚀

**Comment by manthasaigopal** [comment]
Just in case anyone is still confused you can use tools like netron app or torch view to visualize the model. Below you can see the exact detection head of Yolov8s model. 

<img width="2004" height="2186" alt="Image" src="https://github.com/user-attachments/assets/50bdf3d0-c615-4e9b-8dc5-78828030a4f8" />

**Comment by glenn-jocher** [comment]
@saigopalm thanks for sharing that visualization! The netron app is indeed a great tool for understanding model architectures. Your YOLOv8 detection head visualization clearly shows the three detection scales and the separate bbox/cls outputs that @MagiPrince and @CogSP were asking about earlier - this should help clarify the structure for anyone implementing from scratch.

**Comment by manthasaigopal** [comment]
I have doubt in the detection head shown in this image below. They could have concatenated output from DFL directly with class scores right? What is this additional computaion that's done in blue. 

<img width="2004" height="2186" alt="Image" src="https://github.com/user-attachments/assets/19a4468e-616d-433c-83fa-09af45c4da9b" />

**Comment by glenn-jocher** [comment]
@saigopalm that's a great question! The additional computation in blue is crucial because YOLOv8 uses **Distribution Focal Loss (DFL)** for more precise bounding box regression. The 64 channels from DFL represent probability distributions over possible coordinate values (not direct coordinates), so they need to be converted to actual bbox coordinates through a softmax + weighted sum operation. Simply concatenating the raw DFL distributions with class scores wouldn't give you usable coordinates - you'd still need this conversion step during inference. This approach allows YOLOv8 to model uncertainty in bounding box predictions and achieve higher localization accuracy compared to direct coordinate regression.

**Comment by manthasaigopal** [comment]
Given that you do not publish papers on your models can you tell me from which model or paper have you taken this inspiration and followed for the anchorless head?

**Comment by manthasaigopal** [comment]
Can you elaborate more on the computations happening in the enircled region in the image I gave. I want in detail explanation about the detection head. The computation I just mentioned above. How exact coordinates are computed? What is task assigned aligner and why is it used in YOLOv8 ?

**Comment by glenn-jocher** [comment]
@saigopalm great question! The DFL computation converts probability distributions to coordinates like this: each bbox coordinate (x,y,w,h) uses 16 channels representing a probability distribution over possible values. Through softmax + weighted sum with learnable offsets, these distributions become precise coordinates. The TaskAlignedAssigner dynamically assigns ground truth to predictions based on both classification score and IoU alignment - unlike fixed anchor assignment, it considers the task-aligned quality (classification confidence × IoU^α) to find optimal positive samples during training. This anchor-free approach was inspired by FCOS/ATSS papers but adapted specifically for YOLO's multi-scale detection paradigm. The [YOLOv8 output extraction tutorial](https://www.ultralytics.com/blog/extracting-outputs-from-ultralytics-yolov8) covers some implementation details you might find helpful for understanding the coordinate conversion process.
